models/stock_location.py

```diff
@@ -54,12 +97,274 @@ def get_location(self, location_identifier):
             domain = ['|', ('barcode', '=', location_identifier),
                            ('name', '=', location_identifier)]
         else:
-            raise ValidationError(_('Unable to create domain for location search from identifier of type %s') % type(location_identifier))
+            raise ValidationError(
+                _('Unable to create domain for location search from '
+                  'identifier of type %s') % type(location_identifier))
 
         results = self.search(domain)
+
         if not results:
-            raise ValidationError(_('Location not found for identifier %s') % str(location_identifier))
-        if  len(results) > 1:
-            raise ValidationError(_('Too many locations found for identifier %s') % str(location_identifier))
+            raise ValidationError(
+                _('Location not found for identifier %s')
+                % str(location_identifier))
+
+        if len(results) > 1:
+            raise ValidationError(
+                _('Too many locations found for identifier %s')
+                % str(location_identifier))
 
         return results
+
+    #
+    ## Perpetual Inventory
+    #
+
+    def _check_obj_locations(self, loc_keys, obj):
+        Location = self.env['stock.location']
+
+        for key in loc_keys:
+            loc_id = int(obj[key])
+
+            if not Location.browse(loc_id).exists():
+                raise ValidationError(_("The request has an unknown location, "
+                                        "id: '%d'.") % loc_id)
+
+    def _validate_perpetual_inventory_request(self, request):
+        keys = ['location_id', 'location_dest_id']
+
+        if PI_COUNT_MOVES in request:
+            for obj in request[PI_COUNT_MOVES]:
+                self._check_obj_locations(keys, obj)
+
+        if PRECEDING_INVENTORY_ADJUSTMENTS in request:
+            if not request.get(INVENTORY_ADJUSTMENTS):
+                raise ValidationError(
+                    _('You must specify inventory adjustments if you require '
+                      'preceding adjustments.'))
+
+            self._check_obj_locations(keys[:1], request[INVENTORY_ADJUSTMENTS])
```

Comment:
We should also check the location of the preceding inventory adjustments. Also when there are no preceding inventory adjustments but there is an inventory adjustment the location is not checked.

Reply:
Hmm, not sure I'm following here and I might be missing something.

We're checking the preceding inv adjs; `test03_validate_pi_request_failure_no_adj` covers the case you're talking about.
And there's no location in the `inventory_adjustments` items.
Mind that the outer `location_id` (the mandatory entry of `pi_request`) is already checked, otherwise we won't be here.

Reply:
Ops, my bad, wrong token.